common/chains/client/node.go

```diff
@@ -0,0 +1,441 @@
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/assets"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+const queryTimeout = 10 * time.Second
+
+var errInvalidChainID = errors.New("invalid chain id")
+
+var (
+	promPoolRPCNodeVerifies = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_verifies",
+		Help: "The total number of chain ID verifications for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodeVerifiesFailed = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_verifies_failed",
+		Help: "The total number of failed chain ID verifications for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+	promPoolRPCNodeVerifiesSuccess = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: "pool_rpc_node_verifies_success",
+		Help: "The total number of successful chain ID verifications for the given RPC node",
+	}, []string{"chainID", "nodeName"})
+)
+
+type rawclient struct {
+	uri url.URL
+}
+
+type Node[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any, // event filter query options
+	TX_RECEIPT types.ChainReceipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+] interface {
+	RPCClient[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
+	// State returns NodeState
+	State() NodeState
+	// StateAndLatest returns NodeState with the latest received block number & total difficulty.
+	StateAndLatest() (NodeState, int64, *utils.Big)
+	// Name is a unique identifier for this node.
+	Name() string
+	String() string
+
+	Order() int32
+	Start(context.Context) error
+	Close() error
+}
+
+type node[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any, // event filter query options
+	TX_RECEIPT types.ChainReceipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+	SUB types.Subscription,
+] struct {
+	utils.StartStopOnce
+	lfcLog              logger.Logger
+	rpcLog              logger.Logger
+	name                string
+	id                  int32
+	chainID             CHAIN_ID
+	nodePoolCfg         types.NodePool
+	noNewHeadsThreshold time.Duration
+	order               int32
+
+	ws   rawclient
+	http *rawclient
+
+	rpcClient ChainRPCClient[CHAIN_ID, SEQ, ADDR, BLOCK_HASH, TX, TX_HASH, EVENT, EVENT_OPS, TX_RECEIPT, FEE, HEAD, SUB]
```

Comment:
In order to be able to use the multi-node failover code with any chain, we must not tangle it up with a generic client. These are not the same thing, and the client generics are polluting what can be a very simple failover abstraction.
```go
type node[RPC SomeMinimalInterface] struct {
...
rpcClient RPC
...
}

type SomeMinimalInterface interface {
// only what is absolutely required by the failover logic
// no receipts, fees, events, txes, or anything else
}
```
The generic `ChainRPCClient` may be useful in its own right, but it must not be used (directly) in the failover logic.

Does this make sense? It is critical for non-EVM chains.

Reply:
+1